vllm/v1/cudagraph_dispatcher.py

```diff
@@ -0,0 +1,121 @@
+# SPDX-License-Identifier: Apache-2.0
+# SPDX-FileCopyrightText: Copyright contributors to the vLLM project
+from typing import Optional
+
+from vllm.config import CompilationLevel, CUDAGraphMode, VllmConfig
+from vllm.forward_context import BatchDescriptor
+from vllm.logger import init_logger
+from vllm.v1.attention.backends.utils import AttentionCGSupport
+
+logger = init_logger(__name__)
+
+
+class CudagraphDispatcher:
+    """
+    Runtime cudagraph dispatcher to switch between multiple cudagraphs.
+    """
+
+    def __init__(self, vllm_config: VllmConfig):
+        self.vllm_config = vllm_config
+        self.compilation_config = vllm_config.compilation_config
+        self.cudagraph_mode = self.compilation_config.cudagraph_mode
+
+        # Dict to store valid cudagraph dispatching keys.
+        self.cudagraph_keys: dict[CUDAGraphMode, set[BatchDescriptor]] = {
+            CUDAGraphMode.PIECEWISE: set(),
+            CUDAGraphMode.FULL: set(),
+        }
+
+        # Verify if correctly piecewise compilation for attention.
+        piecewise_compilation = not vllm_config.model_config.enforce_eager\
+            and self.compilation_config.level == CompilationLevel.PIECEWISE
+        self.piecewise_attn_compilation = piecewise_compilation and\
+            self.compilation_config.is_attention_splitting
+
+        self.keys_initialized = False
+
+    def add_cudagraph_key(self, runtime_mode: CUDAGraphMode,
+                          batch_descriptor: BatchDescriptor):
+        assert runtime_mode in [CUDAGraphMode.PIECEWISE, CUDAGraphMode.FULL], \
+            f"Invalid cudagraph runtime mode: {runtime_mode}"
+        self.cudagraph_keys[runtime_mode].add(batch_descriptor)
+
+    def initialize_cudagraph_keys(self, capture_mixed_batches: bool,
+                                  uniform_decode_query_len: int,
+                                  attn_cg_support: AttentionCGSupport):
+        # This should be called only after attention backend is initialized.
+
+        # Note: we create all valid keys possible for cudagraph but do not
+        # guarantee all keys would be used. For example, we create keys for
+        # piecewise cudagraphs when it is piecewise compilation, which is always
+        # valid, but for attention backend support unified routine, we may not
+        # trigger capturing/replaying the piecewise cudagraphs depending on
+        # CompilationConfig.cudagraph_mode. In addition, if we allow lazy
+        # capturing in future PR, some keys may never be triggered.
+
+        if self.compilation_config.level == CompilationLevel.PIECEWISE\
+                and len(self.compilation_config.splitting_ops)>0:
```

Comment:
Here I would also check whether splitting_ops include the attention ops or not; if not and we're splitting on some different op, we might be assuming piecewise ignores attention when it doesn't. In the future, this might be a feature where there's another op we want to avoid (not necessarily attention) and then we should treat piecewise cudagraphs just like full (which your PR does a great job of).